### PR TITLE
 #236: implement it tests for reservation

### PR DESCRIPTION
**Pull request by vladarefiev**
PR for https://github.com/timelesslounge/timelessis/issues/236:
- created all it tests for reservations
- refactored old tests
- removed old useless tests

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#435](https://github.com/timelesslounge/timelessis/issues/435) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#435](https://github.com/timelesslounge/timelessis/issues/435) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/c0c9810b-317d-4bb1-8ec5-43b17ea3ae76); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by vladarefiev** [comment]
@amihaiemil Take a look pls

File: tests/factories.py [hunk 1]
```diff
@@ -101,3 +102,16 @@ class Meta:
         model = restaurant_models.Floor
         sqlalchemy_session = DB.session
         sqlalchemy_session_persistence = "commit"
+
+
+class ReservationFactory(factory.alchemy.SQLAlchemyModelFactory):
+    num_of_persons = factory.Faker("pyint")
+    comment = factory.Faker("text")
+    start_time = factory.Faker("date")
+    end_time = factory.Faker("date")
```
Comment (amihaiemil) [inline tests/factories.py]: @vladarefiev I think we shouldn't use ``Faker`` for start/end dates because it might choose start >= end, which wouldn't be correct, right?

File: tests/integration/it_reservations_test.py [hunk 2]
```diff
@@ -1,42 +1,45 @@
...
-        "num_of_persons": 2,
-        "comment": comment,
-        "status": ReservationStatus.confirmed.name
-    })
-    assert response.location.endswith(url_for("reservations.list_reservations"))
+from tests import factories
+from timeless.restaurants.models import Reservation
+
+
+@pytest.mark.skip
```
Comment (amihaiemil) [inline tests/integration/it_reservations_test.py]: @vladarefiev can you add the reason for skipping this test? I think you can do it by adding paranthesys: ``("reason for skipping")``

File: tests/integration/it_reservations_test.py [hunk 3]
```diff
@@ -1,42 +1,45 @@
...
+    assert response.status_code == HTTPStatus.OK
+    assert reservation.comment in response.data
+
+
+def test_create(client):
+    reservation = factories.ReservationFactory.build()
+    reservation_dict = {
+        column.name: str(getattr(reservation, column.name))
+        for column in reservation.__table__.columns
+    }
```
Comment (amihaiemil) [inline tests/integration/it_reservations_test.py]: @vladarefiev I've no idea what the code above is doing, is there some simpler/more understandable way of writing the same logic?

File: tests/integration/it_reservations_test.py [hunk 4]
```diff
@@ -1,42 +1,45 @@
...
-        comment="test comment"
-    )
-    db_session.add(reservation)
-    db_session.commit()
-    response = client.post(url_for('reservations.delete', id=id))
-    assert response.location.endswith(url_for("reservations.list_reservations"))
-    assert Reservation.query.filter_by(id=id).count() == 0
+
+
+@pytest.mark.skip
```
Comment (amihaiemil) [inline tests/integration/it_reservations_test.py]: @vladarefiev reason here as well pls

File: tests/integration/it_reservations_test.py [hunk 5]
```diff
@@ -1,42 +1,45 @@
...
+    reservation_new = factories.ReservationFactory()
+    url = url_for("reservations.edit", id=reservation_old.id)
+    response = client.post(url, data=reservation_new)
+    assert response.status_code == HTTPStatus.OK
+    assert reservation_new.comment in response.data
+
+
+def test_delete(client):
+    reservation = factories.ReservationFactory()
+    url = url_for("reservations.delete", id=reservation.id)
```
Comment (amihaiemil) [inline tests/integration/it_reservations_test.py]: @vladarefiev one time variable, can you inline the call bellow?

File: timeless/restaurants/models.py [hunk 6]
```diff
@@ -133,11 +133,6 @@ class Reservation(TimestampsMixin, DB.Model):
 
     tables = DB.relationship("TableReservation", back_populates="reservation")
 
-    @validate_required("start_time", "end_time", "num_of_persons", "comment",
-                       "status")
-    def __init__(self, **kwargs):
-        super(Reservation, self).__init__(**kwargs)
-
```
Comment (amihaiemil) [inline timeless/restaurants/models.py]: @vladarefiev why did you remove the validation rules? :D

File: tests/factories.py (hunk 1 above)
Comment (vladarefiev) [reply tests/factories.py]: @amidatonga You are correct, but it's possible if  needed to do like that:
`factories.ReservationFactory(end_date=..., ...)`

File: tests/integration/it_reservations_test.py (hunk 2 above)
Comment (vladarefiev) [reply tests/integration/it_reservations_test.py]: @amihaiemil okay

File: tests/integration/it_reservations_test.py (hunk 4 above)
Comment (vladarefiev) [reply tests/integration/it_reservations_test.py]: @amihaiemil +

File: tests/integration/it_reservations_test.py (hunk 5 above)
Comment (vladarefiev) [reply tests/integration/it_reservations_test.py]: @amihaiemil I don't like how it looks like `url = url_for("reservations.delete", id=reservation
factories.ReservationFactory().id)`

if you are not totally against, I will leave ot :)

File: timeless/restaurants/models.py (hunk 6 above)
Comment (vladarefiev) [reply timeless/restaurants/models.py]: @amihaiemil It's useless, we dont need it anywhere,  forms and factories does not work with it.
Db models can validate it by themself, and we should validate it in forms, if we use forms and templates.

File: tests/integration/it_reservations_test.py (hunk 3 above)
Comment (vladarefiev) [reply tests/integration/it_reservations_test.py]: @amihaiemil I've just added a comment. Hope it will be helpful for team.

**Comment by vladarefiev** [comment]
@amihaiemil Thanks for comments, fixed, Take a look pls

File: tests/factories.py (hunk 1 above)
Comment (amihaiemil) [reply tests/factories.py]: @vladarefiev yes, but won't your tests fail from time to time, inexpectedlly? If this test sets start > end, then it will fail when inserting into the DB, I think.

**Comment by amihaiemil** [comment]
@vladarefiev looks ok, just one more question about those faked dates :)

File: tests/factories.py (hunk 1 above)
Comment (vladarefiev) [reply tests/factories.py]: @amihaiemil done. I found a compromise ;)

Comment (vladarefiev) [reply tests/factories.py]: @amihaiemil by default start_time is now, end_time = now + 3 days. But if you want you can set customs dates.

**Comment by vladarefiev** [comment]
@amihaiemil done, pls check

**Comment by amihaiemil** [comment]
@vladarefiev nice, thanks!

**Comment by amihaiemil** [comment]
@rultor merge it

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 19 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Job [#435](https://github.com/timelesslounge/timelessis/issues/435) is not in the agenda of @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), can't inspect

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#435](https://github.com/timelesslounge/timelessis/issues/435) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +25 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)
